x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/PersistentCache.java

```diff
@@ -0,0 +1,616 @@
+        }
+    }
+
+    private static final String CACHE_ID_FIELD = "cache_id";
+    private static final String CACHE_PATH_FIELD = "cache_path";
+    private static final String CACHE_RANGES_FIELD = "cache_ranges";
+    private static final String SNAPSHOT_ID_FIELD = "snapshot_id";
+    private static final String SNAPSHOT_NAME_FIELD = "snapshot_name";
+    private static final String INDEX_ID_FIELD = "index_id";
+    private static final String INDEX_NAME_FIELD = "index_name";
+    private static final String SHARD_INDEX_NAME_FIELD = "shard_index_name";
+    private static final String SHARD_INDEX_ID_FIELD = "shard_index_id";
+    private static final String SHARD_ID_FIELD = "shard_id";
+    private static final String FILE_NAME_FIELD = "file_name";
+    private static final String FILE_LENGTH_FIELD = "file_length";
+
+    private static String buildId(CacheFile cacheFile) {
+        return buildId(cacheFile.getFile());
+    }
+
+    private static String buildId(Path path) {
+        return path.getFileName().toString();
+    }
+
+    private static Term buildTerm(CacheFile cacheFile) {
+        return buildTerm(buildId(cacheFile));
+    }
+
+    private static Term buildTerm(String cacheFileUuid) {
+        return new Term(CACHE_ID_FIELD, cacheFileUuid);
+    }
+
+    private static Document buildDocument(NodeEnvironment.NodePath nodePath, CacheFile cacheFile, SortedSet<Tuple<Long, Long>> cacheRanges)
+        throws IOException {
+        final Document document = new Document();
+        document.add(new StringField(CACHE_ID_FIELD, buildId(cacheFile), Field.Store.YES));
+        document.add(new StringField(CACHE_PATH_FIELD, nodePath.indicesPath.relativize(cacheFile.getFile()).toString(), Field.Store.YES));
+
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            output.writeVInt(cacheRanges.size());
+            for (Tuple<Long, Long> cacheRange : cacheRanges) {
+                output.writeVLong(cacheRange.v1());
+                output.writeVLong(cacheRange.v2());
+            }
+            output.flush();
+            document.add(new StoredField(CACHE_RANGES_FIELD, output.bytes().toBytesRef()));
+        }
+
+        final CacheKey cacheKey = cacheFile.getCacheKey();
+        document.add(new StringField(FILE_NAME_FIELD, cacheKey.getFileName(), Field.Store.YES));
+        document.add(new StringField(FILE_LENGTH_FIELD, Long.toString(cacheFile.getLength()), Field.Store.YES));
+
+        final SnapshotId snapshotId = cacheKey.getSnapshotId();
+        document.add(new StringField(SNAPSHOT_NAME_FIELD, snapshotId.getName(), Field.Store.YES));
+        document.add(new StringField(SNAPSHOT_ID_FIELD, snapshotId.getUUID(), Field.Store.YES));
+
+        final IndexId indexId = cacheKey.getIndexId();
+        document.add(new StringField(INDEX_NAME_FIELD, indexId.getName(), Field.Store.YES));
+        document.add(new StringField(INDEX_ID_FIELD, indexId.getId(), Field.Store.YES));
```

Comment:
Yes, those ones are referring to the index in the snapshot while shard's index name/id are referring to the index assigned to the data node.

Reply:
Maybe one point question here:

Should we simplify `CacheKey` maybe before we move to persisting the format? like we're going to here? There is no need to store the `indexId` IMO. The following information is enough to uniquely identify a shard in a given snapshot:
* snapshot_uuid
* index name
* shard id (`int`)

All the other information like index uuid and and `IndexId` are just noise. Having redundant `index_uuid` in here also technically allows for having the same file in the cache twice for different mounts of an index? (not that it makes sense to have those but it complicates the logic for figuring out what is already in the cache for a given index in the allocation work)

Reply:
We can indeed simplify the CacheKey and removes snapshot's name and snapshot's index uuid. We need to keep the shard's IndexId because as of today cache files are located within a specified shard data path, so they belong to a specific shard within the cluster and need to follow this shard lifecycle.

Reply:
> because as of today cache files are located within a specified shard data path, so they belong to a specific shard within the cluster and need to follow this shard lifecycle

Makes sense thanks :) shall we drop that here or is BwC trivial if we chose to simplify this later (it looks trivial actually and we can just ignore the extra fields in the documents in a follow-up ... but maybe I'm missing something)?

Reply:
You're welcome :) I think that the change will be trivial as you said. Actually I think I can do it as a follow up but because this is spread in many classes doing it now will introduce a lot of noise (and potential merge conflicts).